### PR TITLE
refactor(audio): persist Output Device by stable ma_device_id, not enumeration index (#427)

### LINKED ISSUE #427: refactor(audio): persist Output Device by stable ma_device_id, not enumeration index
## Context

Sub-issue of #375. Follow-up to #417 and #425.

### How the selected Output Device is stored today

- The PI (`sdpi-select setting="audioOutputDevice" global datasource="getAudioDevices"`) binds `audioOutputDevice` in plugin-global settings to the **enumeration index** of the selected miniaudio device — a stringified integer like `"0"`, `"1"`, …, or `"-1"` for System Default.
- On startup, `packages/iracing-plugin-stream-deck/src/plugin.ts` reads the saved integer and calls `native.setAudioDevice(index)`.
- The native layer (`packages/audio-native/src/addon.cc:SetAudioDevice`) enumerates devices via `ma_context_get_devices` and opens the one at that index.

### Why this is fragile

miniaudio's enumeration index is **ephemeral**: it's the position in the current call's return array, not a stable identifier. The order can shift when:

- A device is plugged or unplugged between sessions.
- A driver reinitialises (e.g., Windows Audio Session restart).
- Different USB ports on the same headset / DAC.
- OS audio-preferences changes (e.g., setting a different "Default Device").

If "HyperX Cloud" was index 2 when saved and is index 3 on the next launch, the plugin opens the wrong device. From the user's perspective: "my audio moved itself to the speakers again".

### The stable identifier that already exists

`ma_device_info.id` is miniaudio's platform-native stable device ID:

- **Windows WASAPI** — the endpoint ID, an MMDevice-style GUID-string that survives reboots and PnP events.
- **macOS CoreAudio** — the persistent `AudioDeviceID` / UID.
- Other backends expose equivalent stable IDs.

`packages/audio-native/src/addon.cc:SetAudioDevice` already resolves the index to `pPlaybackDevices[deviceIndex].id` and passes that ID to `ma_engine_init`. The ID is the authoritative identifier at the native layer — it just isn't surfaced or persisted yet.

### 1. Expose `ma_device_id` across the audio-native boundary

`packages/audio-native/src/addon.cc`:

- Add a stable string representation to `ma_device_info` in `GetAudioDevices()` — for example, a base64 / hex encoding of the raw `ma_device_id` bytes. The encoding format itself is opaque to JS; it just needs to round-trip losslessly into `SetAudioDeviceById`.
- Add a new native function `SetAudioDeviceById(deviceId: string): boolean` that decodes the string, enumerates devices, looks up the matching `ma_device_info.id`, and opens it. Keep `SetAudioDevice(index)` for backward compatibility.
- Emit the same device-id string from `GetAudioDevices()` as a new field: `{ index, name, id: string, isDefault }`. (Keep `index` for backward compatibility and for legacy paths that still reference it.)

`packages/audio-native/src/index.ts` and `packages/audio-service/src/audio-service.ts`:

- Mirror the extended `AudioDeviceInfo` type.
- Add `setAudioDeviceById(id: string): boolean`.

### 2. Persist the stable ID, not the index

`packages/iracing-plugin-stream-deck/src/plugin.ts` — the `getAudioDevices` datasource handler added in #425:

```ts
const items = [
  { label: "System Default", value: "" },  // empty string = system default, unambiguous
  ...devices.map(d => ({
    label: `${d.name}${d.isDefault ? " (Default)" : ""}`,
    value: d.id,                           // stable id, not index
  })),
];
```

The legacy `-1` sentinel is replaced with the empty string so the stored type is uniformly `string` (no more mixed "integer-or-negative-sentinel").

### 3. Migrate the apply-device path

Also in `plugin.ts`:

```ts
let currentAudioDeviceId: string | null = null;
onGlobalSettingsChange((settings) => {
  const s = settings as Record<string, unknown>;
  const saved = s.audioOutputDevice;
  const deviceId = typeof saved === "string" ? saved : "";

  if (deviceId === currentAudioDeviceId) return;

  currentAudioDeviceId = deviceId;

  if (deviceId === "") {
    getAudio().setAudioDevice(-1);              // system default
  } else {
    getAudio().setAudioDeviceById(deviceId);    // stable id → re-resolves per session
  }
});
```

### 4. Handle legacy empty-or-absent values

On first launch after the change, `audioOutputDevice` might be empty / missing / a pre-#425 numeric string. Treat any non-matching stored value as System Default at startup — no silent rewrite, no migration code. The project is pre-v1 with a single user (the maintainer); re-picking the device once post-upgrade is acceptable.

### 5. Mid-session device-list changes (best-effort)

miniaudio supports a device notification callback (`ma_context_config.notificationCallback` / `ma_device_notification_proc`). When the system enumeration changes (device plug / unplug), the plugin should:

- Re-enumerate devices.
- If the saved `audioOutputDevice` id still exists → re-apply it to the audio engine (handles re-plug of a previously selected device).
- If the id no longer exists → stay with whatever miniaudio fell back to; do **not** overwrite the persisted setting so the next plug restores selection.

If miniaudio's callback proves awkward to surface through the Node addon, fall back to polling `getAudioDevices()` on a modest interval (e.g., every 10 s) when any Pit Crew PI is open — the scope of this issue is "survives mid-session changes if feasible"; coarse polling is acceptable.

### 6. Mirabox parity

The stub `sendToPropertyInspector` / `onSendToPlugin` no-ops on the VSD adapter remain correct. The new `setAudioDeviceById` path is a no-op on Mirabox since the Pit Crew action isn't shipped there (audio feature-flagged off).

### 7. Tests

- `packages/audio-native/src/addon.test.ts` (if present) / `mock-impl.ts`:
  - Mock `getAudioDevices()` returns items with stable `id` fields.
  - Mock `setAudioDeviceById(id)` returns `true` for known IDs, `false` for unknown.
- `packages/audio-service/src/audio-service.test.ts`: mirror tests for the new method.
- Add a unit test for legacy/empty values: stored `""`, `null`, a stale numeric string, or any non-matching id all fall back to System Default without throwing.

### 8. Documentation sync

- `docs/plugins/core/actions/pit-crew.md` and `packages/website/.../pit-crew.md` — note that Output Device selection survives device list changes.
- `.claude/rules/sdpi-components.md` (if the datasource example references a device index) — point to the id-based pattern.

## Acceptance

- [ ] `audioOutputDevice` is persisted as the miniaudio stable `ma_device_id` string (empty string for System Default), not an enumeration index.
- [ ] Selecting a device in the PI, closing Stream Deck, replugging an unrelated device so indices shift, relaunching — iRaceDeck still opens the originally-selected device.
- [ ] Legacy / empty / unknown stored values fall back to System Default without throwing (no migration code needed; pre-v1, single user).
- [ ] **Best-effort:** device plug/unplug mid-session restores the selected device automatically when it reappears. Not a blocker for this issue if miniaudio's notification path proves prohibitive — fall back to periodic re-resolution documented in the commit body.
- [ ] `pnpm -w build`, `pnpm -w lint`, `pnpm -w test`, `pnpm format` all green.

## Commit message

`refactor(audio): persist Output Device by stable ma_device_id, not enumeration index (#<this issue>)`

## Depends on

- #425 — datasource switch (in progress). This issue lands on top.

Sub-issue of #375.

### PR DESCRIPTION
**Pull request by niklam**
## Related Issue

Fixes #427

## What changed?

Sub-issue of #375. Follow-up to #417.

The miniaudio enumeration index is volatile across sessions: adding, removing, or replugging an unrelated device shifts every other index, so a saved index reopens the wrong device on the next launch. `ma_device_info.id` is the platform-stable identifier (WASAPI endpoint GUID on Windows, CoreAudio UID on macOS) — switch persistence to that.

**Native** (`packages/audio-native/src/addon.cc`):
- Hex-encode `ma_device_id` (full `sizeof(ma_device_id)` bytes) so the raw layout round-trips losslessly through a JS string.
- Add `id` to every `getAudioDevices()` entry alongside `index`/`name`/`isDefault`. `index` is retained for backward compat with `setAudioDevice(index)`.
- Add `setAudioDeviceById(idStr)`: decodes the hex, looks up the device by raw-bytes match against the live enumeration, reinitializes the engine. On failure falls back to system default so the mixer stays usable.

**Service + mock**:
- Mirror the new field and method on `AudioNative` and `AudioNativeMock` (mock honors a synthetic `mock-device-0` id and rejects others).
- Add `setAudioDeviceById` to `IAudioService` and `AudioService` with the same channel-cleanup pattern as `setAudioDevice`. Returns `false` on unknown id; caller (the plugin) decides recovery.

**PI component** (`packages/pi-components/src/components/audio-device-select.ts`):
- Switch option `value` from index string to the stable id string. The System Default sentinel is now the empty string (was `\"-1\"`), so the persisted setting type is uniformly a string id.
- Skip device records missing `id`, and reject empty-string id to avoid colliding with the System Default option.
- Treat unknown / legacy / numeric-index persisted values as missing — the dropdown falls back to System Default and the user re-picks once.

**Plugin glue** (both Stream Deck and Mirabox plugins):
- Persist as id string; fall back to system default on stale or unknown id without rewriting the persisted setting (so a replugged device re-binds automatically once it reappears in the enumeration).
- Skip the redundant default-open call on the very first settings receipt (engine already opened the system default during `init()`).

**Docs** (audio-native/CLAUDE.md, action docs, website docs) updated to reflect the id-based contract.

## Known limitation

Mid-session device-list change (best-effort acceptance criterion in the issue) is **deferred** — neither a notification callback nor polling re-resolution is wired up. Mid-session unplug/replug requires a plugin restart to re-bind. Documented in code; can be added in a follow-up issue.

## Dependency / merge order

Depends on #425 conceptually (the in-flight \`sdpi-select datasource\` refactor). When #425 lands:
- The legacy \`ird-audio-device-select\` component this PR modified is deleted by #425.
- The new datasource handler in #425 emits \`value: d.id\` (the same field this PR adds to the enumeration).

Both PRs end at the same correct state. Either order works; minor merge resolution required when stacking.

## How to test

- \`pnpm test\` — 3187 tests pass (added: mock id field, new \`setAudioDeviceById\` paths, PI legacy-index fallback).
- \`pnpm lint\` / \`pnpm format\` — clean.
- \`pnpm build\` (Stream Deck + Mirabox + native addon) — clean.
- Manual:
  1. Open the Pit Crew Property Inspector, select a non-default Output Device.
  2. Restart Stream Deck. The selected device must still be selected and audible.
  3. Unplug the device, restart Stream Deck. The dropdown must show System Default; the persisted setting must be unchanged (verify in plugin global settings).
  4. Replug the device, restart Stream Deck. The originally-selected device must be selected and audible again — no re-pick required.

## Checklist

- [x] Linked to an approved issue
- [x] New code has unit tests
- [x] Changes registered in all applicable plugins (Stream Deck, Mirabox)
- [x] No unrelated changes included

## Summary by CodeRabbit

* **New Features**
  * Output device selection now persists using platform-stable device identifiers so chosen outputs are reapplied reliably across sessions.

* **Bug Fixes**
  * Prevents loss of audio output selection when devices are reordered, disconnected/reconnected, or when OS audio preferences change.

* **Documentation**
  * Docs updated to explain stable-identifier persistence for the Output Device setting.

### CODE REVIEW

File: packages/audio-service/src/audio-service.ts [hunk 1]
```diff
@@ -359,9 +367,34 @@ class AudioService implements IAudioService {
...
+    if (ok) {
+      this.logger.info("Audio output device switched by id");
+      this.logger.debug(`Device id: ${deviceId}`);
     } else {
-      this.logger.error(`Failed to switch audio output device to index ${deviceIndex}`);
+      // A stale or unknown id — caller decides how to recover (typically
+      // fall back to system default via setAudioDevice(-1)). Log at warn
+      // since it's expected after a device unplug, not an internal failure.
+      this.logger.warn("Audio output device id not found in current enumeration");
+      this.logger.debug(`Unknown device id: ${deviceId}`);
```
[the remark this answers is not in the document]
Comment (niklam) [reply packages/audio-service/src/audio-service.ts]: Good catch — fixed in 1bbaaceb. The failure-path now re-checks the enumeration via `getAudioDevices().some()` to distinguish "id not in current enumeration" (warn — expected after a device unplug) from "id is present but engine reinit failed" (error — internal failure). The extra native call only fires on the failure path, so the cost is negligible.

**Comment by niklam** [comment]
@coderabbitai re: the nitpick about extracting a shared engine-switch helper between `SetAudioDevice` and `SetAudioDeviceById` (`packages/audio-native/src/addon.cc:649-692`):

I considered this and decided to leave the duplication for now. The two paths share the teardown / reinit / fallback shape but diverge in pre-conditions (one validates an integer index against the enumeration count and looks up the device id by index; the other decodes a hex string and finds the device by `memcmp`). The "skip if same as current" optimization in `SetAudioDevice` keys off `g_selectedDeviceIndex`, which `SetAudioDeviceById` doesn't currently track, so a clean shared helper would require either threading "kind: index | id" through the helper or dropping the skip optimization in the index path.

It's ~30 lines of duplication that I'd rather keep visible than hide behind a helper that has to know about both code paths' state. Happy to revisit if a third caller appears, or if a future change actually benefits from the shared shape.

**Comment by niklam** [comment]
@coderabbitai re: the optional debug log in `plugin.ts` when `setAudioDeviceById` returns false:

The audio-service layer already logs that fallback case with full diagnostic context (commit 1bbaaceb, in response to your earlier review):

```ts
// packages/audio-service/src/audio-service.ts setAudioDeviceById fallback
const exists = this.native.getAudioDevices().some((d) => d.id === deviceId);
if (exists) {
  this.logger.error("Failed to switch audio output device by id (engine reinit failed)");
} else {
  this.logger.warn("Audio output device id not found in current enumeration");
}
this.logger.debug(`Device id: ${deviceId}`);
```

So a stale-id fallback already produces a `WARN Audio: Audio output device id not found in current enumeration` line plus the id in DEBUG. Adding another entry in `plugin.ts` would just duplicate that — and per `.claude/rules/logging.md` ("Info Level: Event Only, No Parameters"), I'd want any plugin-side log to be a different event, not a re-statement of the same one.

The plugin's role here is the recovery action (`setAudioDevice(-1)`), which isn't itself worth a log line — `setAudioDevice` already logs `INFO Audio: Audio output device switched` for that.

Skipping this one.
